### ISSUE TITLE
RetryConfigMapping is re-evaluated on each try; and at least twice (when no retries are engaged)

### ISSUE THREAD

**Issue by vkostyukov** [line/armeria#4753]
See [this slack thread](https://line-armeria.slack.com/archives/C1NGPBUH2/p1678687047827359) for more details.

TL;DR: 

`RetryConfigMapping` is evaluated [here](https://github.com/line/armeria/blame/2e91b276816369df6c8d0ece39a81b1e63f69f9d/core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java#L289), which is the code path for every try. There is probably not a lot of value in re-evaluating already computed RetryConfig as it may only add to confusion how these complements interact. Specifically, I'm not sure it's a good idea to swap a `RetryConfig` mid-retry (say on attempt 2 out of 5). What does it even mean, given there is a already a logical state built out of original mapping evaluation.

I'm happy to put together a fix for such that a mapping is evaluated at most once per logical request.

**Comment by minwoox** [comment]
> I'm not sure it's a good idea to swap a RetryConfig mid-retry

That's correct. There's no reason to swap a `RetryConfig`.
Thanks a lot for fixing this in advance. 🙇

**Comment by minwoox** [comment]
Had a chat with the team and there're chances that users want to use different `RetryConfig`s depending on the endpoint in an `EndpointGroup`. For example, users might want to return different delay values according to the weight of an endpoint.

So we have two issues left to solve:
- `RetryConfigMapping.mapping(ctx, req)` is called twice on the first request.
- `RetryConfig.maxTotalAttempts()` and `RetryConfig.responseTimeoutMillisForEachAttempt()` from the second `RetryConfig` are ignored. 

We definitely need to fix the first one. For the second one, we can recalculate the state using `maxTotalAttempts` and `responseTimeoutMillisForEachAttempt` on every attempt. What do you think?

**Comment by vkostyukov** [comment]
> For example, users might want to return different delay values according to the weight of an endpoint.

Endpoints have very dynamic nature at the places where concerns like that should matter. Client's destinations change all the time (assuming proper service discovery system is plugged in) and those addresses usually have very little meaning within the application. I'm honestly having a hard time imagining users encoding those concerns as part of their application logic.

**Comment by vkostyukov** [comment]
@minwoox @trustin What do you think about?

I see a couple of options for us here, in the order I prefer them:

- Option 1: Evaluate `RetryConfigMapping` once, store in the context, reuse it between tries.
- Option 2: Extend `RetryConfig` with `requiresPerTryBackoffMapping` and allow users to opt-in for evaluating a retry config mapping on per try basis (default would be `requiresPerTryBackoffMapping == false`).
- Option 3: Same as 2 but with `requiresPerTryBackoffMapping == true`

Anything else we can do?
